### PR TITLE
Fixed #25532 -- Properly redisplayed invalid JSONField form input values.

### PR DESCRIPTION
**Pull request by claudep**
Thanks David Szotten for the report and Tommy Beadle for code inspiration.

### CODE REVIEW

File: django/contrib/postgres/forms/jsonb.py [hunk 1]
```diff
@@ -27,5 +32,15 @@ def to_python(self, value):
                 params={'value': value},
             )
 
-    def prepare_value(self, value):
+    def bound_data(self, data, initial):
+        if self.disabled:
```
Comment (timgraham) [inline django/contrib/postgres/forms/jsonb.py]: Add a test for these lines?

File: django/contrib/postgres/forms/jsonb.py [hunk 2]
```diff
@@ -27,5 +32,15 @@ def to_python(self, value):
...
+            return initial
+        try:
+            return json.loads(data)
+        except ValueError:
+            return InvalidJSONInput(data)
+
+    def prepare_value(self, value, cleaned=True):
```
Comment (timgraham) [inline django/contrib/postgres/forms/jsonb.py]: remove `cleaned=True`

File: tests/postgres_tests/test_json.py [hunk 3]
```diff
@@ -262,3 +264,22 @@ def test_prepare_value(self):
         field = forms.JSONField()
         self.assertEqual(field.prepare_value({'a': 'b'}), '{"a": "b"}')
         self.assertEqual(field.prepare_value(None), 'null')
+        self.assertEqual(field.prepare_value("foo"), '"foo"')
```
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: single quotes for first string?

File: tests/postgres_tests/test_json.py [hunk 4]
```diff
@@ -262,3 +264,22 @@ def test_prepare_value(self):
...
+        When displaying a bound form (typically due to invalid input), the form
+        should not overquote JSONField inputs.
+        """
+        class JsonForm(Form):
+            name = CharField(max_length=2)
+            jfield = forms.JSONField()
+
+        # JSONField input is fine, name is too long
+        form = JsonForm({'name': 'xyz', 'jfield': '["foo"]'})
+        self.assertIn("[&quot;foo&quot;]</textarea>", form.as_p())
```
Comment (timgraham) [inline tests/postgres_tests/test_json.py]: single quotes

**Comment by timgraham** [comment]
I support backporting to 1.9 unless you have any reservation about it.

**Comment by claudep** [comment]
Now that I had to slightly change `BaseForm._clean_fields`, I'm against backporting the whole patch. We could partially backport without the `test_formfield_disabled` and without the `_clean_fields` part though.

**Comment by timgraham** [comment]
Looks good. I'll leave it up to you whether or not to do the partial backport.

**Comment by claudep** [comment]
My suggested release note for 1.9:
```
* Fixed form's ``JSONField`` contents when redisplaying form values after
  failed form validation (:ticket:`25532`).
```
@timgraham Could you check it?

**Comment by timgraham** [comment]
I think it's sufficient to mention "form" once instead of 3 times, e.g.
```
* Fixed ``JSONField`` inadvertently escaping its contents when displaying values after
  failed form validation (:ticket:`25532`).
```
Thanks Claude!

**Pull request closed without merging**